x-pack/plugins/actions/server/builtin_action_types/lib/connector_token_client.ts

```diff
@@ -0,0 +1,194 @@
+ */
+
+import { omitBy, isUndefined } from 'lodash';
+import { ConnectorToken } from '../../types';
+import { EncryptedSavedObjectsClient } from '../../../../encrypted_saved_objects/server';
+import {
+  Logger,
+  SavedObjectsClientContract,
+  SavedObjectsUtils,
+} from '../../../../../../src/core/server';
+import { CONNECTOR_TOKEN_SAVED_OBJECT_TYPE } from '../../constants/saved_objects';
+
+export const MAX_TOKENS_RETURNED = 1;
+
+interface ConstructorOptions {
+  encryptedSavedObjectsClient: EncryptedSavedObjectsClient;
+  unsecuredSavedObjectsClient: SavedObjectsClientContract;
+  logger: Logger;
+}
+
+interface CreateOptions {
+  connectorId: string;
+  token: string;
+  expiresAt: string;
+  tokenType?: string;
+}
+
+export interface UpdateOptions {
+  id: string;
+  token: string;
+  expiresAt: string;
+  tokenType?: string;
+}
+
+export class ConnectorTokenClient {
+  private readonly logger: Logger;
+  private readonly unsecuredSavedObjectsClient: SavedObjectsClientContract;
+  private readonly encryptedSavedObjectsClient: EncryptedSavedObjectsClient;
+
+  constructor({
+    unsecuredSavedObjectsClient,
+    encryptedSavedObjectsClient,
+    logger,
+  }: ConstructorOptions) {
+    this.encryptedSavedObjectsClient = encryptedSavedObjectsClient;
+    this.unsecuredSavedObjectsClient = unsecuredSavedObjectsClient;
+    this.logger = logger;
+  }
+
+  /**
+   * Create new token for connector
+   */
+  public async create({
+    connectorId,
+    token,
+    expiresAt,
+    tokenType,
+  }: CreateOptions): Promise<ConnectorToken> {
+    const id = SavedObjectsUtils.generateId();
+    const createTime = Date.now();
+    try {
+      const result = await this.unsecuredSavedObjectsClient.create(
+        CONNECTOR_TOKEN_SAVED_OBJECT_TYPE,
+        {
+          connectorId,
+          token,
+          expiresAt,
+          tokenType: tokenType ?? 'access_token',
+          createdAt: new Date(createTime).toISOString(),
+        },
+        { id }
+      );
+
+      return result.attributes as ConnectorToken;
+    } catch (err) {
+      this.logger.warn(
+        `Failed to create connector_token for connectorId "${connectorId}" and tokenType: "${
+          tokenType ?? 'access_token'
+        }". Error: ${err.message}`
+      );
+      throw err;
+    }
+  }
+
+  /**
+   * Update connector token
+   */
+  public async update({ id, token, expiresAt, tokenType }: UpdateOptions): Promise<ConnectorToken> {
+    const { attributes, references, version } =
+      await this.unsecuredSavedObjectsClient.get<ConnectorToken>(
+        CONNECTOR_TOKEN_SAVED_OBJECT_TYPE,
+        id
+      );
+    const createTime = Date.now();
+
+    try {
+      const result = await this.unsecuredSavedObjectsClient.create<ConnectorToken>(
+        CONNECTOR_TOKEN_SAVED_OBJECT_TYPE,
+        {
+          ...attributes,
+          token,
+          expiresAt,
+          tokenType: tokenType ?? 'access_token',
+          createdAt: new Date(createTime).toISOString(),
```

Comment:
Should we consider storing an `updatedAt` time in addition to a `createdAt` time? Would it be useful in the future to know how long the token has been around? Not advocating for one way or another, just wondering...